src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs

```diff
@@ -416,7 +416,29 @@ private static bool TryGetRepresentativeTypeNameFromValue(object value, out PSTy
                 value = PSObject.Base(value);
                 if (value != null)
                 {
-                    type = new PSTypeName(value.GetType());
+                    var typeObject = value.GetType();
+                    
+                    if (typeObject.FullName.Equals("System.Management.Automation.PSObject", StringComparison.OrdinalIgnoreCase))
+                    {
+                        var psobjectPropertyList = new List<PSMemberNameAndType>();
+                        foreach (var property in PSObject.AsPSObject(value).Properties)
```

Comment:
It seems `value` is already PSObject. I mean we can do the type check in previous `if` like `if (value is PSObject value1) ...`

Reply:
Objects are sometimes wrapped in psobject wrappers. One example where that happens is if you call a C# cmdlet:

```
PS C:\> $Test = New-TimeSpan
PS C:\> $Test -is [psobject]
True
PS C:\> $Test2 = [timespan]::new(100)
PS C:\> $Test2 -is [psobject]
False
PS C:\>
```

so if I check it without first unwrapping it with `PSObject.Base()` I will get an inaccurate result.

Reply:
My question is not about  `PSObject.Base()`. Why do we use string in the `if` instead of `if (value is PSObject value1)`?

Reply:
I need the `Type` object for the `PSTypeName` constructor in both branches, that's why I'm calling `GetType` before the if/else statement. Because I already have the `Type` object I figure it's faster to do the string comparison than using `is`.

Reply:
I see your point.
It looks a bit expensive if we first unwrap and then wrap in PSObject.

Reply:
`PSObject.Base()` won't unwrap it if it's a PSObject, it just returns the object as it is. `PSObject.AsPSObject()` simply does:

```
            if (obj is PSObject so)
            {
                return so;
            }
```

So I don't think there's a big performance impact here. I guess we could cast it directly to PSObject for a minor performance gain if you want.

Reply:
In the case why do we PSObject.AsPSObject()?

Reply:
Because I didn't know better when I initially wrote it. I will change it to a cast later today.